src/uct/cuda/cuda_copy/cuda_copy_iface.h

```diff
@@ -17,19 +18,21 @@
 typedef uint64_t uct_cuda_copy_iface_addr_t;
 
 
-typedef enum uct_cuda_copy_stream {
-    UCT_CUDA_COPY_STREAM_H2D,
-    UCT_CUDA_COPY_STREAM_D2H,
-    UCT_CUDA_COPY_STREAM_LAST
-} uct_cuda_copy_stream_t;
+typedef struct uct_cuda_copy_queue_desc {
+    cudaStream_t                stream;
+    ucs_queue_head_t            outstanding_event_q;
+    ucs_queue_head_t            *active_queue;
```

Comment:
why needed?

Reply:
We introduced it avoid going through all `outstanding_event_q` for each [src_mem_type][dst_mem_type] combination as part of iface progress to avoid excessive polling. Instead we maintain a list of active queues in `active_queue`

Reply:
the list head is in face. list element is 'queue'. where is '*active_queue' being used?

Reply:
We populate this field to non-null value first [here](https://github.com/openucx/ucx/pull/6908/files#diff-3d338b2219f6061da816a72f3a6cde940f55450c156a252afb3d87a2e550c07fR131-R134)
```C
    if (q_desc->active_queue == NULL) {
        q_desc->active_queue = &iface->active_queue;
        ucs_queue_push(&iface->active_queue, &q_desc->queue);
    }
```
This way, if another operation is enqueued on the stream, we avoid enqueuing q_desc->queue to iface->active_queue.

Later we set q_desc->active_queue to NULL if all operations in the queue have completed [here](https://github.com/openucx/ucx/pull/6908/files#diff-686df63f8c9b48cbd37c55f24fe8f178accd9fa2fc4dd96b1767f5ca2b67f4a0R200-R203):
```C
        if (ucs_queue_is_empty(event_q)) {
            q_desc->active_queue = NULL;
            ucs_queue_del_iter(&iface->active_queue, iter);
        }
```

Reply:
can we push to &iface->active_queue when outstanding_event_q becomes non-empty?

Reply:
@yosefe ucs/datastruct/queue isn't thread safe, right? If so, it seems like we also need make `uct_cuda_copy_post_cuda_async_copy` thread safe.

Reply:
UCT API does not guarantee thread safety.
Are there any internal cuda_copy threads? if not , no need a lock

Reply:
No internal threads. So I won't add a lock.